### PR TITLE
lib/uksignal: Register system call from uksignal

### PR DESCRIPTION
**Pull request by razvanvirtan**
Register the system calls implemented in the uksignal library. The signal
library provides support for the standard signals and do not yet support
real time signals. The `rt_sigtimedwait` system call does not yet support
time out.
Also, remove the uksignal LIBSYSCALL_SHIM dependency.

Signed-off-by: Răzvan Vîrtan <virtanrazvan@gmail.com>
Signed-off-by: Sharan Santhanam <sharan.santhanam@neclab.eu>

### CODE REVIEW

**Review by mogasergiu — APPROVED** [review]
Looks good to me.

Reviewed-by: Sergiu Moga <sergiu.moga@protonmail.com>

**Review by mogasergiu — APPROVED** [review]
Looks good to me.

Reviewed-by: Sergiu Moga <sergiu.moga@protonmail.com>

File: lib/uksignal/signal.c
Comment (marcrittinghaus) [inline lib/uksignal/signal.c]:
```suggestion
	return uk_thread_sigmask(how, set, oldset);
```

File: lib/uksignal/exportsyms.uk
Comment (marcrittinghaus) [inline lib/uksignal/exportsyms.uk]: Look for proper spacing in the files. Here is one empty line too many. It is the same in line 36

**Comment by razvand** [comment]
@marcrittinghaus , see if this is OK with you and approve it.

File: lib/uksignal/signal.c
Comment (marcrittinghaus) [inline lib/uksignal/signal.c]: We already have a config option: `UK_LIBC_SYSCALLS`, see http://docs.unikraft.org/developers-app.html#syscall-shim-layer. Please adopt the outlined scheme:

```C
UK_SYSCALL_R_DEFINE(int, rt_sigaction, ...)
{
   /* Implement syscall */
}

#if UK_LIBC_SYSCALLS
int sigaction(...)
{
      return rt_sigaction(...)
}
#endif /* UK_LIBC_SYSCALLS */
```

For a more stringent version, you can also use `UK_LLSYSCALL_R_DEFINE` and call `uk_syscall_e_rt_sigaction`. This is a bit cleaner from the perspective that `rt_sigaction` is not a defined libc function and this would prevent its definition. But I am also fine with keeping `UK_SYSCALL_R_DEFINE` because for calling the low-level version all parameters must be cast to `long` which is not so nice.

File: lib/uksignal/signal.c
Comment (marcrittinghaus) [inline lib/uksignal/signal.c]: The `rt_sigaction`/`uk_syscall_e_rt_sigaction` already takes care of setting `errno` and the return value as part of the `UK_SYSCALL_R_DEFINE` macro. So this additional code actually breaks the return values everywhere!

File: lib/uksignal/signal.c
Comment (marcrittinghaus) [inline lib/uksignal/signal.c]: Since `uk_signal` is not used elsewhere, please put its code here and remove the function. This is cleaner and also otherwise this gives a warning when libc declarations are deactivated as `uk_signal` is unused then.

File: lib/uksignal/signal.c
Comment (marcrittinghaus) [inline lib/uksignal/signal.c]: You must not call the function that actually implements the syscall here because that would prevent the system call hook to be triggered, for example, when logging system calls. So you have to call `rt_sigprocmask`/`uk_syscall_e_rt_sigprocmask` here.

File: lib/uksignal/signal.c
Comment (marcrittinghaus) [inline lib/uksignal/signal.c]: Use `NULL` instead of `0` for pointer values everywhere

File: lib/nolibc/musl-imported/include/signal.h
Comment (marcrittinghaus) [inline lib/nolibc/musl-imported/include/signal.h]: As these definitions are not used by any of the public functions, these definitions should be removed from the `signal.h`. Also `time_t` and `suseconds_t` is not actually used even internally.

So place something like this directly in `signal.c` in the include section:
```C
#ifndef __NEED_struct_timespec
#define __NEED_struct_timespec
#endif
#include <sys/types.h>
```

File: lib/uksignal/Config.uk
Comment (marcrittinghaus) [inline lib/uksignal/Config.uk]: This setting should be removed. See below.

File: lib/uksignal/Config.uk
Comment (marcrittinghaus) [inline lib/uksignal/Config.uk]: `LIBSYSCALL_SHIM_LIBCSTUBS` is not required and should be removed.

`LIBSYSCALL_SHIM` can also be omitted. For this please replace the `uk_syscall_r_getpid()` call in `killpg` with 
`-pgrp` as suggested by `man killpg`.

File: lib/uksignal/signal.c
Comment (marcrittinghaus) [inline lib/uksignal/signal.c]:
```suggestion
	*info = ptr->wait.received_signal;
```

Then you can remove the `include <string.h>`.

File: lib/uksignal/exportsyms.uk
Comment (marcrittinghaus) [inline lib/uksignal/exportsyms.uk:36]: Is there a reason why `kill` is not exposed as system call?

File: lib/uksignal/exportsyms.uk
Comment (marcrittinghaus) [inline lib/uksignal/exportsyms.uk:43]: Please group these with the other kill calls

**Review by marcrittinghaus — CHANGES_REQUESTED** [review]
Hi @razvanvirtan ,

thank you for the PR. Before we can merge this PR we need to do some changes. Please see the comments.

Also please take care that the individual commits are in good shape. 
* For example, commit 83e5a514ba breaks the definition of `sigpending` which is then repaired in another commit. 
* Remove all changes that are undone anyways like the `__sigaction`-style definitions. Commits should always be as close to the final code as possible and not reflect the development steps.
* Take care that a commit does not do more than what its description and title suggests.

Do you think you can make this until the release tomorrow?

**Comment by marcrittinghaus** [comment]
@mogasergiu 
Please also take a look at the comments when you have time to get an idea of additional things to consider when reviewing 😃 PR quality is still very poor and it should not have gotten an approval.

**Comment by razvanvirtan** [comment]
@marcrittinghaus This is a rebase of an older PR, maybe it would  be better to close it and create a new one? I will try to do this in the next few days, so that we don't postpone it a lot after the release, but also make sure we don't add low quality code into the upstream.

**Comment by marcrittinghaus** [comment]
Hi @razvanvirtan,

As we have the whole discussion here, I would keep the PR open and just do a force push.

File: lib/uksignal/signal.c
Comment (mogasergiu) [inline lib/uksignal/signal.c]: This definition will conflict with the one above if libc wrappers are enabled. I guess you could remove this one here.

File: lib/uksignal/signal.c
Comment (mogasergiu) [inline lib/uksignal/signal.c:354]: What `si`? It does not seem defined to me.
```suggestion
        siginfo_t si;
```

File: lib/uksignal/signal.c
Comment (mogasergiu) [inline lib/uksignal/signal.c]: This issues a big preproccessing error on my end. For whatever reason, it does not like the name `tkill`. If I were to place some random name here, say `gkill`, that error does not occur anymore. Is this just on my end?

**Comment by mogasergiu** [comment]
By the way, I think that, conventionally, there should be two separate `Signed-off-by` tags instead of one with the authors being separated by a comma.

File: lib/uksignal/signal.c
Comment (razvanvirtan) [reply lib/uksignal/signal.c]: Apparently the tkill symbol was not added to UK_PROVIDED_SYSCALLS, sorry for this, I think something was messed up during the rebase.

File: lib/uksignal/signal.c
Comment (marcrittinghaus) [inline lib/uksignal/signal.c]: This function still sets `errno` in lines 398 and 403. This should be:
```C
	if (pid != 1 && pid != 0 && pid != -1)
		return -ESRCH;

	if (!uk_sig_is_valid(sig))
		return -EINVAL;
```

You can also remove the extra empty line 396.

File: lib/uksignal/signal.c
Comment (marcrittinghaus) [inline lib/uksignal/signal.c:211]: Here is an empty line too many.

File: lib/uksignal/signal.c
Comment (marcrittinghaus) [inline lib/uksignal/signal.c]: Please always provide the macro you tested as comment with the `endif`
```suggestion
#endif /* UK_LIBC_SYSCALLS */
```

**Review by marcrittinghaus — COMMENTED** [review]
Hi @razvanvirtan, thanks for updating the PR! I have some more feedback 😏

Do you have some application or code snipped for testing?

**Comment by razvanvirtan** [comment]
@marcrittinghaus I have updated according to your comments. Here you can find a test source file, I changed it so that it shows a `PASSED` message for each working syscall. The `rt_sigaction + rt_kill` scenario is the only one that doesn't show a message, I think this happens because of the `TO DO` item in the code. The output I am getting right now is the following:
```
rt_sigaction + tkill PASSED
rt_sigpending PASSED
rt_sigsuspend PASSED
rt_sigtimedwait PASSED
```
[Test_signal.zip](https://github.com/unikraft/unikraft/files/9354927/Test_signal.zip)

File: lib/uksignal/signal.c
Comment (marcrittinghaus) [inline lib/uksignal/signal.c]: I recognized that the signature is wrong. It should be:
```suggestion
UK_SYSCALL_R_DEFINE(int, tkill, int __unused, tid, int, sig)
```

**Review by marcrittinghaus — COMMENTED** [review]
Hey @razvanvirtan,

I just recognized that the signature for `tkill` is wrong. Otherwise, thanks for putting all the work in! I have tested the code with the `main.c` that you supplied (thanks!) and I get the same output. I also tested the code without the PR and it is basically the same. So it is not induced by the PR 👍🏻

As signals break with the `musl` changes, I guess we have to update the code anyways and can then fix this broken test.

**Comment by razvanvirtan** [comment]
@marcrittinghaus Thank you for looking into this! Shall we then fix this today or after the release? (Indeed, I haven't tested yet with the musl changes)

**Comment by marcrittinghaus** [comment]
> @marcrittinghaus Thank you for looking into this! Shall we then fix this today or after the release? (Indeed, I haven't tested yet with the musl changes)

Please fix the signature today. Everything else, we do after the release. It would also be outside of the scope of this PR.

Actually, I have not tested with musl either, but I know from Simon that signals break. So there will probably be changes necessary.

**Comment by razvanvirtan** [comment]
@marcrittinghaus Done, I have updated the header.

**Review by marcrittinghaus — APPROVED** [review]
@razvanvirtan

Great thanks! Looks good then 😃👍🏻

**Comment by razvanvirtan** [comment]
@marcrittinghaus I did the rebase to remove the merge conflicts. Also updated a bit the commit messages. Now, some commits are not only registering a new syscall, but also unregister other syscalls, that become libc wrappers.

**Comment by razvanvirtan** [comment]
Last update: changed commit message for [3d7c820](https://github.com/unikraft/unikraft/pull/314/commits/3d7c820e20726d375012e0e9cb48f263ff852c11).

**Review by marcrittinghaus — APPROVED** [review]
@razvanvirtan Thank you very much! Tested it again, all fine 😃👍🏻

Approved-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>

**Pull request closed without merging**
